### ISSUE TITLE
`--deterministic` flag should set a random seed for random processes

### ISSUE THREAD

**Issue by kenibrewer** [fenderglass/ABruijn#640]
As discussed in #509 in this [comment](https://github.com/fenderglass/Flye/issues/509#issuecomment-1548945583), Flye still produces different assembly outputs when using the `--deterministic` flag. That is because multi-threading is not the only source of non-determinism. There is also the use of random variables in the process [here](https://github.com/fenderglass/Flye/blob/7198b7da2a5a244352204bb5c6bc68a4d50e08ce/src/sequence/overlap.cpp#L766C62-L766C62).

An improved `--deterministic` flag should set a random seed for these other processes to allow the entire assembly process to occur deterministically. 

A fully deterministic Flye would be very helpful for integrating Flye into bioinformatic pipelines that are tested against reference datasets.

**Comment by mikolmogorov** [comment]
Makes sense, will add this to the list of TODOs for the next release, thanks!

**Comment by riyasj327** [comment]
@mikolmogorov thank you so much for this amazing tool! Just wondering if this is fixed? As of now, using --deterministic will give the best reproducible assembly?

**Comment by mikolmogorov** [comment]
Not at the moment, unfortunately. True determinism is something very hard to achieve and maintain in assembler, since the algorithm is complex and involves multiple completely different stages. Every stage have to be completely deterministic to make the pipeline deterministic. Given that our resources are limited, not a priority right now.

**Comment by jean-baka** [comment]
Hello! The `--deterministic` flag is no longer mentioned in https://github.com/mikolmogorov/Flye/blob/flye/docs/USAGE.md. Does this mean efforts on getting there have been frozen altogether since your message here from October 2024, @mikolmogorov ? Any chance the team will get on that at some point in the near future? If not, perhaps someone should close this issue with a "not done and not planned" message. ;)

Thanks for the great work on Flye, anyway.

**Comment by dajiaohuang** [comment]
The current flye branch still exposes --deterministic only as a single-threaded disjointig option (flye/main.py:672-674 and flye/assembly/assemble.py:43-44). The pipeline still calls unseeded rand() in src/sequence/overlap.cpp:766-767, src/sequence/sequence_container.cpp:327, and src/sequence/alignment.cpp:24; the current --deterministic path does not seed or replace these calls. The issue therefore remains unresolved on the current branch; no code change was made here.

**Comment by rec3141** [comment]
Some findings from making `flye-modules repeat` reproducible (PR to follow shortly):

- The unseeded `rand()` calls are not the source of the variance. Glibc's `rand()` starts from a fixed seed, and the three sites that matter (k-mer bias sampling, chimera sampling, non-ACGT replacement) run in serial code, so they produce the same sequence every run; the fourth only picks a thread id. Seeding them changes nothing observable.
- The variance comes from **address-ordered iteration**: the repeat graph stores nodes in `unordered_set<GraphNode*>` and the resolver/haplotype code uses ~75 `unordered_{set,map}` containers keyed by `GraphNode*` / `GraphEdge*`. `std::hash<T*>` is the address, which differs between runs under ASLR, and several passes make decisions in iteration order. That is also why `--deterministic` (single-threading the disjointig stage) cannot help: the order varies even single-threaded.
- Secondary sources: `groupBySet` returning an `unordered_map` keyed by `SetNode*`, `ReadAligner::alignReads` pushing alignments in thread-completion order, reused per-thread Dijkstra scratch maps, a pointer comparison used as a priority tie-break, and the matched-pair loop in `resolveConnections` iterating an `unordered_map`.

Empirically (two nanopore metagenomes, 12 k and 150 k edges): 2.9.6 gives the same graph *before* repeat resolution on every run, and different `repeat_graph_dump` / `repeat_graph_edges.fasta` / `graph_after_rr.gv` after it. Hashing the pointers by their sequential ids and ordering the handful of other collection points makes the whole `repeat` stage byte-identical across runs and across thread counts, with the pre-resolution graph unchanged from 2.9.6 and the final graph equal to one of 2.9.6's own outcomes. The `assemble` and polishing stages have their own parallel collection points and are not covered by this.

**Comment by dajiaohuang** [comment]
Thanks for the detailed reproduction and correction. This updates my earlier focus on the unseeded rand() calls: the address-ordered iteration through pointer-keyed unordered containers and thread-completion ordering is the stronger explanation for the observed variance. I will treat the rand() sites as non-causal for this issue and wait for the proposed reproducibility PR rather than make an issue-only patch.
